lib/request-handler.js

```diff
@@ -63,94 +63,100 @@ module.exports = function processRequest (options, request, response) {
+                if (typeof(filterResponseHeaders) === 'function') {
+                    responseHeaders = filterResponseHeaders(fragment.attributes, headers);
                 }
 
-                if (primary && shouldWriteHead) {
-                    shouldWriteHead = false;
-                    fragment.on('response', (statusCode, headers) => {
-                        // Map response headers
-                        if (typeof(filterResponseHeaders) === 'function') {
-                            Object.assign(responseHeaders, filterResponseHeaders(fragment.attributes, headers));
-                        }
-
-                        if (headers.location) {
-                            responseHeaders.location = headers.location;
-                        }
-                        // Make resources early discoverable while processing HTML
-                        const preloadAssets = headers.link ? getAssetsToPreload(headers, request) : '';
-                        if (preloadAssets !== '') {
-                            responseHeaders.link = preloadAssets;
-                        }
-                        this.emit('response', request, statusCode, responseHeaders);
-                        response.writeHead(statusCode, responseHeaders);
-                        resultStream
-                            .pipe(contentLengthStream)
-                            .pipe(response);
-                    });
-                    fragment.on('fallback', (err) => {
-                        this.emit('error', request, err);
-                        response.writeHead(500, responseHeaders);
-                        resultStream
-                            .pipe(contentLengthStream)
-                            .pipe(response);
-                    });
-                    fragment.on('error', (err) => {
-                        this.emit('error', request, err);
-                        response.writeHead(500, responseHeaders);
-                        response.end();
-                    });
+                if (headers.location) {
+                    responseHeaders.location = headers.location;
                 }
 
-                return fragment.fetch(request, false);
-            }
 
-            return handleTag(request, tag);
+                // Make resources early discoverable while processing HTML
+                const preloadAssets = headers.link ? getAssetsToPreload(headers, request) : '';
+                if (preloadAssets !== '') {
+                    responseHeaders.link = preloadAssets;
+                }
+                this.emit('response', request, statusCode, responseHeaders);
+
+                response.writeHead(statusCode, responseHeaders);
+                blockStream
+                    .pipe(contentLengthStream)
+                    .pipe(response);
+            });
+
+            fragment.on('fallback', (err) => {
+                this.emit('error', request, err);
+                // TODO: add some responseHeaders?
```

Comment:
why not `responseHeaders` like before? Do you have anything specific in mind? 

Its also configurable through `filterResponseHeaders` function